### PR TITLE
dump: always split TiDB v4.* tables through tidb rowid to save TiDB's memory

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
fix https://github.com/pingcap/dumpling/issues/104 and close https://github.com/pingcap/dumpling/issues/278

### What is changed and how it works?
Split chunks by `PKIsHandle` for TiDB v4.0.*. Make sure TiDB won't OOM through splitting tables by rowids.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
Test dump with v4.0.0. The result is looks expected.

Related changes

 - Need to cherry-pick to the release branch
 
### Release note

- Always split TiDB v4.* tables through tidb rowid to save TiDB's memory.

### CODE REVIEW

File: v4/export/sql_type.go [hunk 1]
```diff
@@ -18,37 +18,42 @@ var (
 )
 
 func initColTypeRowReceiverMap() {
-	for _, s := range dataTypeString {
+	dataTypeStringArr := []string{
```
Comment (lance6716) [inline v4/export/sql_type.go:21]: please add a comment where we collect those types

File: v4/export/sql.go [hunk 2]
```diff
@@ -435,6 +442,7 @@ func ShowMasterStatus(db *sql.Conn) ([]string, error) {
 
 // GetSpecifiedColumnValue get columns' values whose name is equal to columnName
 func GetSpecifiedColumnValue(rows *sql.Rows, columnName string) ([]string, error) {
+	defer rows.Close()
```
Comment (lance6716) [inline v4/export/sql.go:445]: `GetSpecifiedColumnValueAndClose` or let caller close `rows` 🤔

Comment (lichunzhu) [reply v4/export/sql.go:448]: https://github.com/pingcap/dumpling/pull/273/commits/bfaf7dc29249eeb6b12a071c235f9934387f6a58

File: v4/export/sql_type.go (hunk 1 above)
Comment (lichunzhu) [reply v4/export/sql_type.go:22]: https://github.com/pingcap/dumpling/pull/273/commits/bfaf7dc29249eeb6b12a071c235f9934387f6a58

File: v4/export/dump.go [hunk 3]
```diff
@@ -658,6 +709,124 @@ func buildTiDBTableSampleQuery(pkFields []string, dbName, tblName string) string
...
+		} else {
+			pkVals = append(pkVals, []string{pkVal})
+		}
+	}
+
+	return pkVals, err
+}
+
+func extractTiDBRowIDFromDecodedKey(indexField, key string) (string, error) {
+	if p := strings.Index(key, indexField); p != -1 {
```
Comment (lance6716) [inline v4/export/dump.go:823]: do we have a system test, to get a report when TiDB changes the behaviour 😵

File: v4/export/dump.go [hunk 4]
```diff
@@ -658,6 +709,124 @@ func buildTiDBTableSampleQuery(pkFields []string, dbName, tblName string) string
...
+	var (
+		rows      *sql.Rows
+		startKeys []string
+	)
+	const (
+		partitionRegionSQL = "SHOW TABLE `%s`.`%s` PARTITION(`%s`) REGIONS"
+		regionRowKey       = "r_"
+	)
+	logger := tctx.L().With(zap.String("database", dbName), zap.String("table", tableName), zap.String("partition", partition))
+	rows, err = conn.QueryContext(tctx, fmt.Sprintf(partitionRegionSQL, dbName, tableName, partition))
```
Comment (lance6716) [inline v4/export/dump.go:797]: `dbName` and `tableName` not escaped

File: v4/export/sql.go [hunk 5]
```diff
@@ -433,8 +440,12 @@ func ShowMasterStatus(db *sql.Conn) ([]string, error) {
 	return oneRow, nil
 }
 
-// GetSpecifiedColumnValue get columns' values whose name is equal to columnName
-func GetSpecifiedColumnValue(rows *sql.Rows, columnName string) ([]string, error) {
+// GetSpecifiedColumnValueAndClose get columns' values whose name is equal to columnName and close the given rows
+func GetSpecifiedColumnValueAndClose(rows *sql.Rows, columnName string) ([]string, error) {
```
Comment (lance6716) [inline v4/export/sql.go:444]: we might also ToUpper(columnName) when comparing

File: v4/export/dump.go [hunk 6]
```diff
@@ -658,6 +709,124 @@ func buildTiDBTableSampleQuery(pkFields []string, dbName, tblName string) string
...
+	}
+	return
+}
+
+func checkTiDBTableRegionPkFields(pkFields, pkColTypes []string) (err error) {
+	if len(pkFields) != 1 || len(pkColTypes) != 1 {
+		err = errors.Errorf("unsupported primary key for selectTableRegion. pkFields: [%s], pkColTypes: [%s]", strings.Join(pkFields, ", "), strings.Join(pkColTypes, ", "))
+		return
+	}
+	if _, ok := dataTypeNum[pkColTypes[0]]; !ok {
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply v4/export/dump.go:735]: Is there clustered index feature on TiDB v4.0?

[the remark this answers is not in the document]
Comment (lance6716) [reply v4/export/dump.go:735]: No, do you mean v4.0 won't go here?

File: v4/export/dump.go (hunk 3 above)
Comment (lichunzhu) [reply v4/export/dump.go:823]: I have tested TiDB v4.0.0 and v4.0.12. In TiDB v4.x versions, they work well. I can add a test to make sure TiDB won't change this.

File: v4/export/dump.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lance6716) [reply v4/export/dump.go:735]: oh I see, v5.0 won't go here

[the remark this answers is not in the document]
Comment (lichunzhu) [reply v4/export/dump.go:735]: Yep. v4.0 will always have an integer type row index.

File: v4/export/sql.go (hunk 5 above)
Comment (lichunzhu) [reply v4/export/sql.go:444]: addressed in https://github.com/pingcap/dumpling/pull/273/commits/69c1e7c6012590500e0deceed11c876b49c90f19

File: v4/export/sql_type.go [hunk 7]
```diff
@@ -17,38 +17,44 @@ var (
...
+		"CHAR", "NCHAR", "VARCHAR", "NVARCHAR", "CHARACTER", "VARCHARACTER",
+		"TIMESTAMP", "DATETIME", "DATE", "TIME", "YEAR", "SQL_TSI_YEAR",
+		"TEXT", "TINYTEXT", "MEDIUMTEXT", "LONGTEXT",
+		"ENUM", "SET", "JSON",
+	}
+
+	dataTypeNumArr := []string{
+		"INTEGER", "BIGINT", "TINYINT", "SMALLINT", "MEDIUMINT",
+		"INT", "INT1", "INT2", "INT3", "INT8",
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply v4/export/sql_type.go:37]: Update comments in https://github.com/pingcap/dumpling/pull/273/commits/69c1e7c6012590500e0deceed11c876b49c90f19

File: v4/export/sql_type.go [hunk 8]
```diff
@@ -17,38 +17,44 @@ var (
...
 	}
-	for _, s := range dataTypeNum {
+	for _, s := range dataTypeNumArr {
+		dataTypeNum[s] = struct{}{}
 		colTypeRowReceiverMap[s] = SQLTypeNumberMaker
 	}
-	for _, s := range dataTypeBin {
+	for _, s := range dataTypeBinArr {
+		dataTypeBin[s] = struct{}{}
 		colTypeRowReceiverMap[s] = SQLTypeBytesMaker
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply v4/export/sql_type.go:59]: `SQLTypeBytes` will transfer binary data to `base 16` code data, so I think we won't have this problem.
https://github.com/pingcap/dumpling/blob/bff60f8/v4/export/sql_type.go#L284

File: v4/export/sql_type.go [hunk 9]
```diff
@@ -17,38 +17,50 @@ var (
...
+// The second is to be the receiver of select row type, which will use sql.DB's rows.DatabaseTypeName(), which is from
+// https://github.com/go-sql-driver/mysql/blob/v1.5.0/fields.go#L17-97
 func initColTypeRowReceiverMap() {
-	for _, s := range dataTypeString {
+	dataTypeStringArr := []string{
+		"CHAR", "NCHAR", "VARCHAR", "NVARCHAR", "CHARACTER", "VARCHARACTER",
+		"TIMESTAMP", "DATETIME", "DATE", "TIME", "YEAR", "SQL_TSI_YEAR",
+		"TEXT", "TINYTEXT", "MEDIUMTEXT", "LONGTEXT",
+		"ENUM", "SET", "JSON", "NULL", "VAR_STRING",
+		"GEOMETRY", // TODO: support GEOMETRY later
```
Comment (kennytm) [inline v4/export/sql_type.go:32]: move this to `dataTypeBinArr`.

File: v4/export/sql.go [hunk 10]
```diff
@@ -458,7 +470,8 @@ func GetSpecifiedColumnValue(rows *sql.Rows, columnName string) ([]string, error
 			strs = append(strs, oneRow[fieldIndex].String)
 		}
 	}
-	return strs, nil
+	rows.Close()
```
Comment (kennytm) [inline v4/export/sql.go:473]: with that `defer` you're closing `rows` twice 🤔

File: v4/export/dump.go [hunk 11]
```diff
@@ -572,16 +571,76 @@ func (d *Dumper) selectMinAndMaxIntValue(conn *sql.Conn, db, tbl, field string)
...
+	} else {
+		tctx.L().Debug("dumping TiDB tables with TABLE REGIONS",
+			zap.String("database", db), zap.String("table", tbl))
+		var partitions []string
+		partitions, err = GetPartitionNames(conn, db, tbl)
+		if err == nil {
+			if len(partitions) == 0 {
+				handleColNames, handleVals, err = selectTiDBTableRegion(tctx, conn, db, tbl)
+			} else {
+				return d.concurrentDumpTiDBPartitionTables(tctx, conn, meta, taskChan, partitions)
```
Comment (lance6716) [inline v4/export/dump.go:594]: weak review: could we not enter `concurrentDumpTiDBPartitionTables`, instead, return enough data and reuse below `sendConcurrentDumpTiDBTasks`?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: v4/export/dump.go (hunk 11 above)
Comment (lichunzhu) [reply v4/export/dump.go:594]: Maybe we can do that later

**Pull request merged**